File: server/model/activity.go
package model

import "time"

// Activity 活动信息结构体
type Activity struct {
	OrmBase

	CycleDays int64 `json:"cycleDays" gorm:"not null;comment:周期持续天数"`
	// 长期活动的话，定时器在结算上一个周期之后，会自动发起下一个周期的活动
	Always    bool     `json:"always" gorm:"not null;comment:是否为长期活动"`
	Title     string   `json:"title" gorm:"varchar(80);not null;comment:活动标题"`
	Url       string   `json:"url" gorm:"varchar(200);not null;comment:活动介绍网址"`
	TypeCodes []string `json:"typeCodes" gorm:"type:json;serializer:json;not null;comment:活动接受投稿的类型字典码集合"`
}

// ActivityRecord 活动信息记录
type ActivityRecord struct {
	OrmBase

	BeginTime  time.Time `json:"beginTime" gorm:"not null;comment:开始时间"`
	EndTime    time.Time `json:"endTime" gorm:"not null;comment:结束时间"`
	ActivityId string    `json:"activityId" gorm:"not null;comment:活动id"`
	Activity   Activity  `json:"activity" gorm:"foreignKey:ActivityId"`
}

// ActivityType 活动类型枚举
type ActivityType string

const (
	PostActivity   ActivityType = "11"
	MenuActivity   ActivityType = "12"
	RecipeActivity ActivityType = "13"
)

File: client/lib/model/activity.dart
import 'package:client/common/model.dart';
import 'package:client/model/model.dart';
import 'package:client/tool/date.dart';

/*
* 活动信息
* @author wuxubaiyang
* @Time 2022/10/14 14:56
*/
class ActivityModel extends BaseModel with BasePart {
  // 持续周期
  final num cycleTime;

  // 是否为长期活动
  final bool always;

  // 活动标题
  final String title;

  // 介绍网址
  final String url;

  // 支持的活动类型
  final List<String> typeCodes;

  ActivityModel.from(obj)
      : cycleTime = obj?['cycleTime'] ?? 0,
        always = obj?['always'] ?? false,
        title = obj?['title'] ?? '',
        url = obj?['url'] ?? '',
        typeCodes =
            (obj?['typeCodes'] ?? []).map<String>((e) => '$e').toList() {
    initBasePart(obj);
  }

  @override
  Map<String, dynamic> to() => {
        ...basePart,
        'cycleTime': cycleTime,
        'always': always,
        'title': title,
        'url': url,
        'typeCodes': typeCodes,
      };
}

/*
* 活动记录
* @author wuxubaiyang
* @Time 2022/10/14 14:59
*/
class ActivityRecordModel extends BaseModel with BasePart {
  // 开始时间
  final DateTime beginTime;

  // 结束时间
  final DateTime endTime;

  // 活动id
  final String activityId;

  // 活动信息
  final ActivityModel activity;

  ActivityRecordModel.from(obj)
      : beginTime = DateTime.tryParse(obj?['beginTime'] ?? '') ?? DateTime(0),
        endTime = DateTime.tryParse(obj?['endTime'] ?? '') ?? DateTime(0),
        activityId = obj?['activityId'] ?? '',
        activity = ActivityModel.from(obj?['activity'] ?? {}) {
    initBasePart(obj);
  }

  @override
  Map<String, dynamic> to() => {
        ...basePart,
        'beginTime': beginTime.toIso8601StringWithUTC(),
        'endTime': endTime.toIso8601StringWithUTC(),
        'activityId': activityId,
        'activity': activity.to(),
      };
}
